src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -138,7 +138,7 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                     {
                         CreateObject = options.MemberAccessorStrategy.CreateConstructor(type);
 
-                        PropertyInfo[] properties = type.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+                        PropertyInfo[] properties = type.GetProperties(BindingFlags.Instance | BindingFlags.Public);
```

Comment:
We should add some tests ensuring that the serializer doesn't deserialize/serialize non-public fields under any circumstance including:

- when a derived type adds a non-public `new` property which collides with a public property defined on the parent
- when there are property name collisions between public/non-public properties due to `[JsonPropertyName]`
- when there are property name collisions between public/non-public properties due to `[JsonNamingPolicy]`

all such tests should have variants checking that public properties that are `[JsonIgnore]`'d are not accidentally (de)serialized due to mismatching `JsonPropertyInfo` with a colliding non-public member.

https://dotnetfiddle.net/2BD7bK goes over a few expecations.

Reply:
@YohDeadfall - let us know if you are up for adding the tests in this PR based on what @layomia suggested (its fine if you prefer someone else take that on, but will certainly be helpful to add).

Reply:
Though about improving tests, but as a separate PR (would like to do some cleanup and improvements for better coverage/understanding). Will take a look at this tomorrow.